File: js/packages/cyvest-js/src/keys.ts
/**
 * Key generation utilities for Cyvest objects.
 *
 * Provides deterministic, unique key generation for all object types.
 * Keys are used for object identification, retrieval, and merging.
 */

/**
 * Key type prefixes used in Cyvest.
 */
export type KeyType = "obs" | "fnd" | "evd" | "sig" | "rel" | "dec" | "tag";

/** Every prefix a v7 document can carry. `enr:` is gone — enrichments are evidence now. */
export const KEY_PREFIXES: readonly KeyType[] = ["obs", "fnd", "evd", "sig", "rel", "dec", "tag"] as const;

/**
 * Normalize a string value for consistent key generation.
 */
function normalizeValue(value: string): string {
  return value.trim().toLowerCase();
}

function normalizeObservableValue(
  obsType: string,
  value: string,
  subtype?: string
): string {
  const normalizedType = normalizeValue(obsType);
  const normalizedSubtype = subtype ? normalizeValue(subtype) : undefined;
  const stripped = value.trim();

  if (normalizedType === "command_line") return stripped;
  if (normalizedType === "email" || normalizedType === "host") {
    return stripped.toLowerCase();
  }
  if (
    normalizedType === "user" &&
    (normalizedSubtype === "email" || normalizedSubtype === "upn")
  ) {
    return stripped.toLowerCase();
  }
  if (normalizedSubtype === "uid" || normalizedSubtype === "pid") {
    if (!/^[+-]?\d+$/.test(stripped)) {
      throw new Error(`${normalizedSubtype} observable values must be base-10 integers`);
    }
    return BigInt(stripped).toString(10);
  }
  return stripped;
}

function sha256(content: string): string {
  const bytes = new TextEncoder().encode(content);
  const words: number[] = [];
  const bitLength = bytes.length * 8;
  for (const byte of bytes) {
    words.push(byte);
  }
  words.push(0x80);
  while ((words.length % 64) !== 56) words.push(0);
  for (let i = 7; i >= 0; i--) {
    words.push(Math.floor(bitLength / 2 ** (i * 8)) & 0xff);
  }

  const h = [
    0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a,
    0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19,
  ];
  const k = Array.from({ length: 64 }, (_, index) => {
    let primeCount = 0;
    let candidate = 2;
    while (true) {
      let prime = true;
      for (let divisor = 2; divisor * divisor <= candidate; divisor++) {
        if (candidate % divisor === 0) {
          prime = false;
          break;
        }
      }
      if (prime && primeCount++ === index) {
        return Math.floor((Math.cbrt(candidate) % 1) * 2 ** 32) >>> 0;
      }
      candidate++;
    }
  });
  const rotateRight = (value: number, amount: number) =>
    (value >>> amount) | (value << (32 - amount));

  for (let offset = 0; offset < words.length; offset += 64) {
    const schedule = new Array<number>(64);
    for (let i = 0; i < 16; i++) {
      const base = offset + i * 4;
      schedule[i] =
        ((words[base] << 24) |
          (words[base + 1] << 16) |
          (words[base + 2] << 8) |
          words[base + 3]) >>>
        0;
    }
    for (let i = 16; i < 64; i++) {
      const s0 =
        rotateRight(schedule[i - 15], 7) ^
        rotateRight(schedule[i - 15], 18) ^
        (schedule[i - 15] >>> 3);
      const s1 =
        rotateRight(schedule[i - 2], 17) ^
        rotateRight(schedule[i - 2], 19) ^
        (schedule[i - 2] >>> 10);
      schedule[i] = (schedule[i - 16] + s0 + schedule[i - 7] + s1) >>> 0;
    }

    let [a, b, c, d, e, f, g, hh] = h;
    for (let i = 0; i < 64; i++) {
      const s1 = rotateRight(e, 6) ^ rotateRight(e, 11) ^ rotateRight(e, 25);
      const choice = (e & f) ^ (~e & g);
      const temp1 = (hh + s1 + choice + k[i] + schedule[i]) >>> 0;
      const s0 = rotateRight(a, 2) ^ rotateRight(a, 13) ^ rotateRight(a, 22);
      const majority = (a & b) ^ (a & c) ^ (b & c);
      const temp2 = (s0 + majority) >>> 0;
      hh = g;
      g = f;
      f = e;
      e = (d + temp1) >>> 0;
      d = c;
      c = b;
      b = a;
      a = (temp1 + temp2) >>> 0;
    }
    [a, b, c, d, e, f, g, hh].forEach((value, index) => {
      h[index] = (h[index] + value) >>> 0;
    });
  }
  return h.map((value) => value.toString(16).padStart(8, "0")).join("");
}

function encodeKeyPart(value: string, keepSlash = false): string {
  let encoded = encodeURIComponent(value).replace(/[!'()*]/g, (char) =>
    `%${char.charCodeAt(0).toString(16).toUpperCase()}`
  );
  encoded = encoded.replace(/%40/gi, "@");
  if (keepSlash) encoded = encoded.replace(/%2F/gi, "/").replace(/%5C/gi, "\\");
  return encoded;
}

/**
 * Generate a unique key for an observable.
 *
 * Format: obs:{type}:{normalized_value}
 *
 * @param obsType - Type of observable (ip, url, domain, hash, etc.)
 * @param value - Value of the observable
 * @returns Unique observable key
 *
 * @example
 * ```ts
 * generateObservableKey("ipv4", "192.168.1.1")
 * // => "obs:ipv4:192.168.1.1"
 * ```
 */
export function generateObservableKey(
  obsType: string,
  value: string,
  subtype?: string,
  namespace?: string
): string {
  const normalizedType = normalizeValue(obsType);
  const normalizedSubtype = subtype ? normalizeValue(subtype) : undefined;
  const normalizedNamespace = namespace?.trim().toLowerCase() || undefined;
  const normalizedValue = normalizeObservableValue(
    normalizedType,
    value,
    normalizedSubtype
  );
  const identity = JSON.stringify({
    namespace: normalizedNamespace ?? null,
    subtype: normalizedSubtype ?? null,
    type: normalizedType,
    value: normalizedValue,
  });

  if (
    normalizedType === "command_line" ||
    new TextEncoder().encode(identity).length > 128
  ) {
    return `obs:${normalizedType}:sha256:${sha256(identity)}`;
  }
  if (!normalizedSubtype && !normalizedNamespace) {
    return `obs:${normalizedType}:${normalizedValue.toLowerCase()}`;
  }

  const parts = ["obs", encodeKeyPart(normalizedType), normalizedSubtype ? encodeKeyPart(normalizedSubtype) : "_"];
  if (normalizedNamespace) parts.push(encodeKeyPart(normalizedNamespace));
  parts.push(encodeKeyPart(normalizedValue, true));
  return parts.join(":");
}

/**
 * Generate a finding key.
 *
 * Format: `fnd:{rule_id}`, or `fnd:{rule_id}:{external_id}`.
 *
 * A finding is identified by the rule that produced it. Running the same rule on several
 * observables means several findings, so pass an `externalId` to keep them apart — usually the
 * observable key.
 *
 * @example
 * ```ts
 * generateFindingKey("url_in_body")
 * // => "fnd:url_in_body"
 * ```
 */
export function generateFindingKey(ruleId: string, externalId?: string): string {
  const normalizedRule = normalizeValue(ruleId);
  if (externalId) {
    return `fnd:${normalizedRule}:${externalId.trim()}`;
  }
  return `fnd:${normalizedRule}`;
}

/**
 * Generate an observable-signal key.
 *
 * Format: `sig:{source}:{subject_key}`. The prefix names the *family*, not its first member:
 * v6's `ti:` would have locked every future signal kind behind a threat-intel name.
 *
 * @example
 * ```ts
 * generateSignalKey("virustotal", "obs:ipv4:192.168.1.1")
 * // => "sig:virustotal:obs:ipv4:192.168.1.1"
 * ```
 */
export function generateSignalKey(source: string, subjectKey: string, externalId?: string): string {
  const normalizedSource = normalizeValue(source);
  if (externalId) {
    return `sig:${normalizedSource}:${externalId.trim()}:${subjectKey}`;
  }
  return `sig:${normalizedSource}:${subjectKey}`;
}

/** Kept under its v6 name; new code calls {@link generateSignalKey}. */
export const generateThreatIntelKey = generateSignalKey;

/**
 * Generate a relation key.
 *
 * Format: `rel:{kind}:{source_key}>{target_key}` — direction lives in the key itself, so no
 * `direction` field is needed to disambiguate.
 */
export function generateRelationKey(
  sourceKey: string,
  targetKey: string,
  kind: string,
  externalId?: string,
): string {
  const normalizedKind = normalizeValue(kind);
  if (externalId) {
    return `rel:${normalizedKind}:${externalId.trim()}:${sourceKey}>${targetKey}`;
  }
  return `rel:${normalizedKind}:${sourceKey}>${targetKey}`;
}

/**
 * Generate a decision key.
 *
 * Format: `dec:{target_key}` — one decision per target, whatever it says. The kind is content,
 * not identity: a target holds a single current stance, and changing one's mind is a
 * re-assertion the merge law settles by freshness.
 */
export function generateDecisionKey(targetKey: string): string {
  return `dec:${targetKey}`;
}

/**
 * Generate a unique key for a tag.
 *
 * Format: tag:{normalized_name}
 *
 * @param name - Name of the tag (can use : as hierarchy delimiter)
 * @returns Unique tag key
 *
 * @example
 * ```ts
 * generateTagKey("header:auth:dkim")
 * // => "tag:header:auth:dkim"
 * ```
 */
export function generateTagKey(name: string): string {
  const normalizedName = normalizeValue(name);
  return `tag:${normalizedName}`;
}

/**
 * Get all ancestor tag names from a hierarchical tag name.
 *
 * @param name - Tag name with : delimiter
 * @returns Array of ancestor tag names
 *
 * @example
 * ```ts
 * getTagAncestors("header:auth:dkim")
 * // => ["header", "header:auth"]
 * ```
 */
export function getTagAncestors(name: string): string[] {
  const parts = name.split(":");
  const ancestors: string[] = [];
  for (let i = 0; i < parts.length - 1; i++) {
    ancestors.push(parts.slice(0, i + 1).join(":"));
  }
  return ancestors;
}

/**
 * Check if a tag is a direct child of another tag.
 *
 * @param childName - Potential child tag name
 * @param parentName - Potential parent tag name
 * @returns True if childName is a direct child of parentName
 *
 * @example
 * ```ts
 * isTagChildOf("header:auth", "header") // => true
 * isTagChildOf("header:auth:dkim", "header") // => false (grandchild)
 * ```
 */
export function isTagChildOf(childName: string, parentName: string): boolean {
  if (!childName.startsWith(parentName + ":")) {
    return false;
  }
  const remaining = childName.slice(parentName.length + 1);
  return !remaining.includes(":");
}

/**
 * Check if a tag is a descendant of another tag (any depth).
 *
 * @param descendantName - Potential descendant tag name
 * @param ancestorName - Potential ancestor tag name
 * @returns True if descendantName is a descendant of ancestorName
 *
 * @example
 * ```ts
 * isTagDescendantOf("header:auth:dkim", "header") // => true
 * isTagDescendantOf("header", "header") // => false (same)
 * ```
 */
export function isTagDescendantOf(descendantName: string, ancestorName: string): boolean {
  return descendantName.startsWith(ancestorName + ":");
}

/**
 * Extract the type prefix from a key.
 *
 * @param key - The key to parse
 * @returns Type prefix (see {@link KEY_PREFIXES}) or null if invalid
 *
 * @example
 * ```ts
 * parseKeyType("obs:ipv4:192.168.1.1") // => "obs"
 * parseKeyType("invalid") // => null
 * ```
 */
export function parseKeyType(key: string): KeyType | null {
  if (key.includes(":")) {
    const prefix = key.split(":", 1)[0] as KeyType;
    if ((KEY_PREFIXES as readonly string[]).includes(prefix)) {
      return prefix;
    }
  }
  return null;
}

/**
 * Validate a key format and optionally finding its type.
 *
 * @param key - The key to validate
 * @param expectedType - Optional expected type prefix
 * @returns True if valid, false otherwise
 *
 * @example
 * ```ts
 * validateKey("obs:ipv4:192.168.1.1") // => true
 * validateKey("obs:ipv4:192.168.1.1", "obs") // => true
 * validateKey("obs:ipv4:192.168.1.1", "chk") // => false
 * validateKey("invalid") // => false
 * ```
 */
export function validateKey(key: string, expectedType?: KeyType): boolean {
  if (!key || !key.includes(":")) {
    return false;
  }

  const keyType = parseKeyType(key);
  if (!keyType) {
    return false;
  }

  if (expectedType && keyType !== expectedType) {
    return false;
  }

  return true;
}

/**
 * Extract components from an observable key.
 *
 * Only reliable for the simple form `obs:{type}:{value}`. A key carrying a subtype or a
 * namespace, and a `command_line` or oversized key folded into `obs:{type}:sha256:{digest}`,
 * both return the remaining segments verbatim as `value` — a key is an identity token, not a
 * record. Read the observable from the document instead of parsing its key.
 *
 * @param key - Observable key to parse
 * @returns Object with type and value, or null if invalid
 *
 * @example
 * ```ts
 * parseObservableKey("obs:ipv4:192.168.1.1")
 * // => { type: "ipv4", value: "192.168.1.1" }
 * ```
 */
export function parseObservableKey(
  key: string
): { type: string; value: string } | null {
  if (!validateKey(key, "obs")) {
    return null;
  }
  const parts = key.split(":");
  if (parts.length >= 3) {
    return {
      type: parts[1],
      value: parts.slice(2).join(":"), // Handle values with colons
    };
  }
  return null;
}

File: src/cyvest/keys.py
"""
Key generation utilities for Cyvest objects.

Provides deterministic, unique key generation for all object types.
Keys are used for object identification, retrieval, and merging.
"""

import hashlib
import json
from typing import Any
from urllib.parse import quote

from cyvest.enums import ObservableSubtype, ObservableType

_MAX_READABLE_OBSERVABLE_IDENTITY_BYTES = 128


def _normalize_value(value: str) -> str:
    """
    Normalize a string value for consistent key generation.

    Args:
        value: The value to normalize

    Returns:
        Normalized lowercase string
    """
    return value.strip().lower()


def _hash_dict(data: dict[str, Any]) -> str:
    """
    Create a deterministic hash from a dictionary.

    Args:
        data: Dictionary to hash

    Returns:
        SHA256 hash of the sorted dictionary items
    """
    # Sort keys for deterministic ordering
    sorted_items = sorted(data.items())
    content = str(sorted_items)
    return hashlib.sha256(content.encode()).hexdigest()[:16]


def slugify(value: str, *, fallback: str = "") -> str:
    """
    Lower-case kebab-case identifier from free text: ``"URL In Body!"`` → ``"url-in-body"``.

    For a ``rule_id`` or a tag derived from a label a model or a human wrote. Returns ``fallback``
    when nothing survives.
    """
    normalized = "".join(ch.lower() if ch.isalnum() else "-" for ch in value)
    collapsed = "-".join(part for part in normalized.split("-") if part)
    return collapsed or fallback


def normalize_observable_value(obs_type: str, value: str, subtype: str | None = None) -> str:
    """Normalize an observable value for identity without changing its display value."""
    normalized_type = _normalize_value(obs_type)
    normalized_subtype = _normalize_value(subtype) if subtype else None
    stripped = value.strip()

    if normalized_type == ObservableType.COMMAND_LINE.value:
        return stripped
    if normalized_type in {ObservableType.EMAIL.value, ObservableType.HOST.value}:
        return stripped.lower()
    if normalized_type == ObservableType.USER.value and normalized_subtype in {
        ObservableSubtype.USER_EMAIL.value,
        ObservableSubtype.USER_UPN.value,
    }:
        return stripped.lower()
    if normalized_subtype in {
        ObservableSubtype.USER_UID.value,
        ObservableSubtype.PROCESS_PID.value,
    }:
        try:
            return str(int(stripped, 10))
        except ValueError as exc:
            raise ValueError(f"{normalized_subtype} observable values must be base-10 integers") from exc
    return stripped


def generate_observable_key(
    obs_type: str,
    value: str,
    subtype: str | None = None,
    namespace: str | None = None,
) -> str:
    """
    Generate a unique key for an observable.

    Format: obs:{type}:{normalized_value}

    Args:
        obs_type: Type of observable (ipv4, ipv6, url, domain, hash, email, etc.)
        value: Value of the observable

    Returns:
        Unique observable key
    """
    normalized_type = _normalize_value(obs_type)
    normalized_subtype = _normalize_value(subtype) if subtype else None
    normalized_namespace = namespace.strip().lower() if namespace else None
    normalized_value = normalize_observable_value(normalized_type, value, normalized_subtype)
    identity = json.dumps(
        {
            "namespace": normalized_namespace,
            "subtype": normalized_subtype,
            "type": normalized_type,
            "value": normalized_value,
        },
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    )
    if (
        normalized_type == ObservableType.COMMAND_LINE.value
        or len(identity.encode("utf-8")) > _MAX_READABLE_OBSERVABLE_IDENTITY_BYTES
    ):
        digest = hashlib.sha256(identity.encode("utf-8")).hexdigest()
        return f"obs:{normalized_type}:sha256:{digest}"

    if normalized_subtype is None and normalized_namespace is None:
        return f"obs:{normalized_type}:{normalized_value.lower()}"

    parts = ["obs", quote(normalized_type, safe="._-")]
    parts.append(quote(normalized_subtype, safe="._-") if normalized_subtype else "_")
    if normalized_namespace is not None:
        parts.append(quote(normalized_namespace, safe="._-@"))
    parts.append(quote(normalized_value, safe="._-@/\\"))
    return ":".join(parts)


def resolve_observable_key(*args: Any, **kwargs: Any) -> str:
    """
    Accept either a ready-made key or the identity components, and return a key.

    Shared by the facade and the shared context so a worker does not have to remember which of
    the two it is holding.
    """
    if len(args) == 1 and not kwargs and isinstance(args[0], str) and args[0].startswith("obs:"):
        return args[0]

    obs_type = kwargs.get("obs_type", args[0] if args else None)
    value = kwargs.get("value", args[1] if len(args) > 1 else None)
    subtype = kwargs.get("subtype", args[2] if len(args) > 2 else None)
    namespace = kwargs.get("namespace", args[3] if len(args) > 3 else None)
    return generate_observable_key(
        obs_type.value if hasattr(obs_type, "value") else str(obs_type),
        str(value),
        subtype=subtype.value if hasattr(subtype, "value") else subtype,
        namespace=namespace,
    )


def generate_finding_key(rule_id: str, external_id: str | None = None) -> str:
    """
    Generate a finding key.

    Format: ``fnd:{rule_id}`` or ``fnd:{rule_id}:{external_id}``

    A finding is identified by the rule that produced it. Running the same rule on several
    observables means several findings, so pass an ``external_id`` to keep them apart — usually
    the observable key.
    """
    normalized_rule = _normalize_value(rule_id)
    if external_id:
        return f"fnd:{normalized_rule}:{external_id.strip()}"
    return f"fnd:{normalized_rule}"


def generate_relation_key(
    source_key: str,
    target_key: str,
    kind: str,
    external_id: str | None = None,
) -> str:
    """
    Generate a relation key.

    Format: ``rel:{kind}:{source_key}>{target_key}``

    Direction is carried by the key itself, so no ``direction`` field is needed to disambiguate.
    """
    normalized_kind = _normalize_value(kind)
    if external_id:
        return f"rel:{normalized_kind}:{external_id.strip()}:{source_key}>{target_key}"
    return f"rel:{normalized_kind}:{source_key}>{target_key}"


def generate_decision_key(target_key: str) -> str:
    """
    Generate a decision key.

    Format: ``dec:{target_key}``

    **One decision per target**, whatever it says. The kind is content, not identity: a target
    holds a single current stance, and changing one's mind is a re-assertion the merge law
    settles by freshness — not a second fact competing with the first.

    Keying on the kind as well would let ``UPHOLD`` and ``REFUTE`` coexist under distinct keys
    and force the engine to arbitrate them itself, duplicating the very law the store already
    applies.
    """
    return f"dec:{target_key}"


def generate_evidence_key(
    *,
    source: str,
    external_id: str | None,
    evidence_type: str,
    content: Any = None,
    uri: str | None = None,
) -> str:
    """Generate a deterministic evidence key."""
    normalized_source = _normalize_value(source)
    if external_id:
        normalized_external_id = external_id.strip()
        readable = f"evd:{quote(normalized_source, safe='._-')}:{quote(normalized_external_id, safe='._-@/')}"
        if len(readable.encode("utf-8")) <= 128:
            return readable
        digest = hashlib.sha256(f"{normalized_source}\0{normalized_external_id}".encode()).hexdigest()
        return f"evd:{normalized_source}:sha256:{digest}"

    payload = json.dumps(
        {
            "content": content,
            "source": normalized_source,
            "type": evidence_type.strip().lower(),
            "uri": uri,
        },
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        default=str,
    )
    return f"evd:sha256:{hashlib.sha256(payload.encode()).hexdigest()}"


def generate_signal_key(source: str, subject_key: str, external_id: str | None = None) -> str:
    """
    Generate an observable-signal key.

    Format: ``sig:{source}:{subject_key}``. The prefix names the *family*, not its first member:
    v6's ``ti:`` would have locked every future signal kind behind a threat-intel name.

    Passing an ``external_id`` is how a caller opts into keeping a source's history instead of
    letting the freshest assertion win.
    """
    normalized_source = _normalize_value(source)
    if external_id:
        return f"sig:{normalized_source}:{external_id.strip()}:{subject_key}"
    return f"sig:{normalized_source}:{subject_key}"


# Kept under its v6 name for the migration path; new code calls generate_signal_key.
generate_threat_intel_key = generate_signal_key


def generate_tag_key(name: str) -> str:
    """
    Generate a unique key for a tag.

    Format: tag:{normalized_name}

    Args:
        name: Name of the tag (uses : as hierarchy delimiter)

    Returns:
        Unique tag key
    """
    normalized_name = _normalize_value(name)
    return f"tag:{normalized_name}"


def get_tag_ancestors(name: str) -> list[str]:
    """
    Get all ancestor tag names from a hierarchical tag name.

    Uses ":" as the hierarchy delimiter.

    Args:
        name: Tag name (e.g., "header:auth:dkim")

    Returns:
        List of ancestor names (e.g., ["header", "header:auth"])
    """
    parts = name.split(":")
    return [":".join(parts[: i + 1]) for i in range(len(parts) - 1)]


def is_tag_child_of(child_name: str, parent_name: str) -> bool:
    """
    Check if a tag is a direct child of another tag.

    Args:
        child_name: Potential child tag name
        parent_name: Potential parent tag name

    Returns:
        True if child_name is a direct child of parent_name
    """
    if not child_name.startswith(parent_name + ":"):
        return False
    remaining = child_name[len(parent_name) + 1 :]
    return ":" not in remaining


def is_tag_descendant_of(descendant_name: str, ancestor_name: str) -> bool:
    """
    Check if a tag is a descendant (child, grandchild, etc.) of another tag.

    Args:
        descendant_name: Potential descendant tag name
        ancestor_name: Potential ancestor tag name

    Returns:
        True if descendant_name is a descendant of ancestor_name
    """
    return descendant_name.startswith(ancestor_name + ":")


KEY_PREFIXES = frozenset({"obs", "fnd", "evd", "sig", "rel", "dec", "tag"})


def parse_key_type(key: str) -> str | None:
    """
    Extract the type prefix from a key.

    Args:
        key: The key to parse

    Returns:
        Type prefix (one of :data:`KEY_PREFIXES`) or None if invalid
    """
    if ":" in key:
        return key.split(":", 1)[0]
    return None


def parse_observable_key(key: str) -> tuple[str, str] | None:
    """
    Parse an observable key into its type and value.

    Format: obs:{type}:{normalized_value}

    Args:
        key: Observable key to parse

    Returns:
        Tuple of (observable type, value) or None if invalid
    """
    if parse_key_type(key) != "obs":
        return None

    parts = key.split(":", 2)
    if len(parts) != 3:
        return None

    _, obs_type, value = parts
    if not obs_type or not value:
        return None

    return obs_type, value


def validate_key(key: str, expected_type: str | None = None) -> bool:
    """
    Validate a key format and optionally finding its type.

    Args:
        key: The key to validate
        expected_type: Optional expected type prefix

    Returns:
        True if valid, False otherwise
    """
    if not key or ":" not in key:
        return False

    key_type = parse_key_type(key)
    if key_type not in KEY_PREFIXES:
        return False

    if expected_type and key_type != expected_type:
        return False

    return True
